src/librustc/ty/fold.rs

```diff
@@ -162,6 +163,35 @@ pub trait TypeFolder<'gcx: 'tcx, 'tcx> : Sized {
         t.super_fold_with(self)
     }
 
+    #[inline]
+    /// If `false` - the default - then `ty::Invariant` might be used instead of the
+    /// correct variance when folding an item with a variance.
+
+    /// Otherwise, the correct variance is looked up from the tcx, which can
+    /// be a performance and cycle hazard.
+    fn use_variances(&self) -> bool {
```

Comment:
This seems *ok*, but also error-prone. i.e., if somebody forgets to make `use_variances` be true, we'll have a problem. It might be worth thinking if we can turn that into a dynamic assertion or something. But I don't obviously see how to do that with this setup as in this PR -- I was imagining that if the variance was tracked in a field, then accessing it could assert that `use_variances` is true or something...

Reply:
I *could* make `variances` an `Option`, but t hen the users would need to handle that.